pipeline/lib/metric.ml

```diff
@@ -0,0 +1,230 @@
+  let metrics =
+    Array.fold_left
+      (fun acc datum ->
+        match datum with
+        | _commit, benchmark_name, test_name, metrics ->
+            metrics
+            |> List.fold_left
+                 (fun acc
+                      ({ name; value; _ } : Current_bench_json.Latest.metric) ->
+                   let key = (benchmark_name, test_name, name) in
+                   BenchmarksData.add key value acc)
+                 acc)
+      BenchmarksData.empty data
+  in
+  let commit =
+    if Array.length data > 0
+    then
+      let commit, _, _, _ = data.(0) in
+      Some commit
+    else None
+  in
+  (metrics, commit)
+
+let sum_float_list xs = List.fold_left (fun acc x -> acc +. x) 0. xs
+
+let mean_float_list xs =
+  let n = List.length xs in
+  match n with 0 -> 0. | _ -> sum_float_list xs /. Float.of_int n
+
+let avg_of_value (v : Current_bench_json.Latest.value) =
+  match v with
+  | Float v -> v
+  | Floats vs -> mean_float_list vs
+  | Assoc vs -> (
+      vs
+      |> List.filter_map (function
+           | key, value when key = "avg" -> Some value
+           | _ -> None)
+      |> function
+      | [ avg ] -> avg
+      | _ -> 0.)
+
+let calc_diff v cv =
+  if cv <> 0.
+  then (v -. cv) /. cv *. 100.
+  else if v <> 0.
+  then (v -. cv) /. v *. 100.
+  else 0.
+
+let find_changed_metrics ~metrics ~compare_metrics =
+  BenchmarksData.merge
+    (fun _ (value : Current_bench_json.Latest.value option)
+         (compare_value : Current_bench_json.Latest.value option) ->
+      match (value, compare_value) with
+      | Some value, Some compare_value ->
+          let delta =
+            calc_diff (avg_of_value value) (avg_of_value compare_value)
+          in
+          if Float.abs delta > change_threshold then Some delta else None
+      | Some _, None -> Some Float.nan
+      | _ -> None)
+    metrics compare_metrics
+
+module SeenBenchmarks = Set.Make (MetricKey)
+
+let format_changes ~changes =
+  let text_list, _ =
+    BenchmarksData.fold
+      (fun (bench_name, test_name, metric_name) change (acc, seen) ->
+        let bench_key = (bench_name, "", "") in
+        let test_key = (bench_name, test_name, "") in
+        let bench_seen = SeenBenchmarks.mem bench_key seen in
+        let test_seen = SeenBenchmarks.mem test_key seen in
+        let entry, seen =
+          match bench_seen with
+          | false ->
+              ( Fmt.str "\n\n## Benchmark: %s\n" bench_name,
+                SeenBenchmarks.add bench_key seen )
+          | true -> ("", seen)
+        in
+        let entry, seen =
+          match test_seen with
+          | false ->
+              ( Fmt.str "%s\n### Test: %s\n" entry test_name,
+                SeenBenchmarks.add test_key seen )
+          | true -> (entry, seen)
+        in
+        let change =
+          if Float.is_nan change
+          then Fmt.str "- %s is a new metric" metric_name
+          else Fmt.str "- %s changed by %f%%" metric_name change
```

Comment:
What do you think of `%.0f` or `%.1f` ? (not sure that we need super precise percentages)